src/RequestHandler/outofprocess/forwardinghandler.cpp

```diff
@@ -1252,27 +1321,41 @@ None
     if (TlsGetValue(g_dwTlsIndex) != this)
     {
         DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == NULL);
+        if (m_RequestStatus != FORWARDER_RECEIVED_WEBSOCKET_RESPONSE)
+        {
+            // Webscoket has already been guarded by critical section
+            // Only require exclisive lock for non-websocket scenario which has duplex channel
+            // Otherwise, there will be a deadlock
+            AcquireSRWLockExclusive(&m_RequestLock);
+            TlsSetValue(g_dwTlsIndex, this);
+            fExclusiveLocked = TRUE;
+            DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == this);
+        }
+        else
+        {
+            AcquireSRWLockShared(&m_RequestLock);
+            TlsSetValue(g_dwTlsIndex, this);
+            fSharedLocked = TRUE;
+            DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == this);
+        }
+    }
 
-        AcquireSRWLockShared(&m_RequestLock);
-        TlsSetValue(g_dwTlsIndex, this);
-        fLockAcquired = TRUE;
-        DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == this);
+    if (fEndRequest)
```

Comment:
Consider to give a clearer name for this flag - hard to tell the difference between fEndRequest and m_fFinishRequest until realized it just indicated winhttp handle closing.

Reply:
good suggestion. Changed to fHandleClosing